crates/flashblocks-archiver/src/archiver.rs

```diff
@@ -0,0 +1,144 @@
+use crate::database::Database;
+use crate::metrics::Metrics;
+use crate::websocket::WebSocketPool;
+use crate::{config::Config, FlashblockMessage};
+use anyhow::Result;
+use std::collections::HashMap;
+use std::time::{Duration, Instant};
+use tokio::time::interval;
+use tracing::{error, info, warn};
+use uuid::Uuid;
+
+#[derive(Debug)]
+pub struct FlashblocksArchiver {
+    config: Config,
+    database: Database,
+    builder_ids: HashMap<String, Uuid>,
+    metrics: Metrics,
+}
+
+impl FlashblocksArchiver {
+    pub async fn new(config: Config) -> Result<Self> {
+        let database = Database::new(&config.database.url, config.database.max_connections).await?;
+
+        // Run database migrations
+        database.run_migrations().await?;
+
+        // Register builders and get their IDs
+        let mut builder_ids = HashMap::new();
+        for builder_config in &config.builders {
+            let builder_id = database
+                .get_or_create_builder(builder_config.url.as_ref(), Some(&builder_config.name))
+                .await?;
+            builder_ids.insert(builder_config.name.clone(), builder_id);
+        }
+
+        Ok(Self {
+            config,
+            database,
+            builder_ids,
+            metrics: Metrics::default(),
+        })
+    }
+
+    pub async fn run(&self) -> Result<()> {
+        info!(
+            "Starting FlashblocksArchiver with {} builders",
+            self.config.builders.len()
+        );
+
+        if self.config.builders.is_empty() {
+            warn!("No builders configured, archiver will not collect any data");
+            return Ok(());
+        }
+
+        // Start WebSocket connections
+        let ws_pool = WebSocketPool::new(self.config.builders.clone());
+        let mut receiver = ws_pool.start().await?;
+
+        // Set up batching
+        let mut batch = Vec::with_capacity(self.config.archiver.batch_size);
+        let mut flush_interval = interval(Duration::from_secs(
+            self.config.archiver.flush_interval_seconds,
+        ));
+
+        info!("FlashblocksArchiver started, listening for flashblock messages");
+
+        loop {
+            tokio::select! {
+                // Receive flashblock messages
+                message = receiver.recv() => {
+                    match message {
+                        Some((builder_name, payload)) => {
+                            if let Some(builder_id) = self.builder_ids.get(&builder_name) {
+                                batch.push((*builder_id, payload));
+
+                                // Flush if batch is full
+                                if batch.len() >= self.config.archiver.batch_size {
+                                    if let Err(e) = self.flush_batch(&mut batch).await {
```

Comment:
I wonder if we could do this without blocking the routine. Depending on how long flushes take vs. batch size), we could potentially drop data here